library/java/org/chromium/net/impl/CancelProofEnvoyStream.java

```diff
@@ -0,0 +1,114 @@
+package org.chromium.net.impl;
+
+import java.nio.ByteBuffer;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicInteger;
+import io.envoyproxy.envoymobile.engine.EnvoyHTTPStream;
+
+/**
+ * Consistency layer above the {@link EnvoyHTTPStream} preventing unwarranted Stream operations
+ * after a "cancel" operation. There are no "synchronized" - this is Compare And Swap based logic.
+ *
+ * <p>This contraption ensures that once a "cancel" operation is invoked, there will be no further
+ * operations allowed with the EnvoyHTTPStream - subsequent operations will be ignored silently.
+ * However, in the event that that one or more EnvoyHTTPStream operations are currently being
+ * executed, the "cancel" operation gets postponed: the last concurrent operation will invoke
+ * "cancel" at the end.
+ *
+ * <p>Instance of this class start with a state of "Busy Starting". This ensure that if a cancel
+ * is invoked while the stream is being created, that cancel will be executed only once the stream
+ * is completely initialized. Doing otherwise leads to unpredictable outcomes.
+ */
+final class CancelProofEnvoyStream {
+
+  private static final int CANCEL_BIT = 0x8000;
+  private final AtomicInteger mState = new AtomicInteger(1); // Busy starting.
+  private volatile EnvoyHTTPStream mStream;                  // Cancel can come from any Thread.
+
+  /** Sets the stream. Can only be invoked once. */
+  void setStream(EnvoyHTTPStream stream) {
+    assert mStream == null;
+    mStream = stream;
+    if (cancelNeedsToBeInvoked()) {
+      mStream.cancel(); // Cancel was called meanwhile, so now this is honored.
+    }
+  }
+
+  /** Initiates the sending of the request headers if the state permits. */
+  void sendHeaders(Map<String, List<String>> envoyRequestHeaders, boolean endStream) {
+    if (cancelHasAlreadyBeenInvoked()) {
+      return; // Already Cancelled - to late to send something.
+    }
+    mStream.sendHeaders(envoyRequestHeaders, endStream);
+    if (cancelNeedsToBeInvoked()) {
+      mStream.cancel(); // Cancel was called meanwhile, so now this is honored.
+    }
+  }
+
+  /** Initiates the sending of one chunk of the request body if the state permits. */
+  void sendData(ByteBuffer buffer, boolean finalChunk) {
+    if (cancelHasAlreadyBeenInvoked()) {
+      return; // Already Cancelled - to late to send something.
+    }
+    // The Envoy Mobile library only cares about the capacity - must use the correct ByteBuffer
```

Comment:
Yes, EnvoyMobile does not care about buffer.position.

https://github.com/envoyproxy/envoy-mobile/blob/main/library/common/jni/jni_utility.cc#L167

I added the "length" feature - only capacity was considered before...

Reply:
I see. This still sounds super scary. Can you file an issue about this confusion between what the Java layer thinks is in the buffer and what the C layer thinks is in the buffer? You clearly don't need to fix this but it would be good to have this written down and we can sort it out later.

Reply:
Issue #2247 filed.